File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0;
        int maxLength = 0;
        Set<Character> seen = new HashSet();

        for (int right = 0; right < s.length(); right ++) {
            char curr = s.charAt(right);

            while (seen.contains(curr)) {
                seen.remove(s.charAt(left));
                left ++;
            }

            seen.add(curr);
            maxLength = Math.max(maxLength, right - left + 1);

        }   
        return maxLength;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen = set()
        left = 0
        maxLength = 0

        for right in range(len(s)):
            curr = s[right]
            while curr in seen:
                seen.remove(s[left])
                left += 1
            
            seen.add(curr)
            maxLength = max(maxLength, right - left + 1)

        return maxLength
        